### PR TITLE
ADD: additional metric tests covering overlapping sequences

### PR DESCRIPTION
**Pull request by benleetownsend**

### CODE REVIEW

File: finetune/util/metrics.py [hunk 1]
```diff
@@ -36,7 +36,7 @@ def _convert_to_token_list(annotations, doc_idx=None):
 
 def sequence_labeling_token_confusion(text, true, predicted):
     none_class = "<None>"
-    unique_classes = list(set([seq['label'] for seqs in true for seq in seqs]))
+    unique_classes = list(set([seq['label'] for seqs in true + predicted for seq in seqs]))
```
Comment (lhz1029) [inline finetune/util/metrics.py:39]: :+1:

File: tests/test_metrics.py [hunk 2]
```diff
@@ -115,13 +115,13 @@ def check_metrics(self, Y, Y_pred, expected, span_type=None):
         macro_f1 = sequence_f1(Y, Y_pred, span_type=span_type, average="macro")
         for cls_ in counts:
             for metric in counts[cls_]:
-                self.assertEqual(len(counts[cls_][metric]), expected[cls_][metric])
-            self.assertAlmostEqual(recalls[cls_], expected[cls_]["recall"], places=3)
+                self.assertEqual(len(counts[cls_][metric]), expected[cls_][metric], msg=f"class={cls_}, metric={metric}")
+            self.assertAlmostEqual(recalls[cls_], expected[cls_]["recall"], places=3, msg=f"class={cls_}")
```
Comment (lhz1029) [inline tests/test_metrics.py:119]: Why can't this be equal?

File: tests/test_metrics.py [hunk 3]
```diff
@@ -335,3 +335,203 @@ def test_seq_mixed(self):
...
+                "recall": 1.0,
+                "f1-score": 1.0,
+            },
+            "class2": {
+                "false_positives": 1,
+                "false_negatives": 0,
+                "true_positives": 0,
+                "precision": 0.0,
+                "recall": 0.0,
+                "f1-score": 0.0,
```
Comment (lhz1029) [inline tests/test_metrics.py:379]: can we add support?

File: tests/test_metrics.py [hunk 4]
```diff
@@ -335,3 +335,203 @@ def test_seq_mixed(self):
...
+            },
+            "micro-f1": 0.66666, # Calculated as the harmonic mean of Recall = 1, Precision = 0.5
```
Comment (lhz1029) [inline tests/test_metrics.py:381]: why do we get .5 for precision but 1 for recall?

File: tests/test_metrics.py [hunk 5]
```diff
@@ -335,3 +335,203 @@ def test_seq_mixed(self):
...
+                "true_positives": 1,
+                "precision": 1.0,
+                "recall": 1.0,
+                "f1-score": 1.0,
+            },
+            "micro-f1": 1.0,
+            "macro-f1": 1.0,
+            "weighted-f1": 1.0,
+        }
+        for span_type in ["overlap", "superset"]:
```
Comment (lhz1029) [inline tests/test_metrics.py]: I think superset should be different. Namely, recall should be 0 for superset

**Review by lhz1029 — COMMENTED** [review]
All of these tests make me think that it would be very useful to have a confusion matrix in sequence metrics. I think that'll be much easier to interpret for our users in most cases when they want to drill in.

File: finetune/util/metrics.py [hunk 6]
```diff
@@ -281,25 +296,26 @@ def sequence_labeling_counts(true, predicted, equality_fn):
         for annotations in [true_annotations, predicted_annotations]:
             for annotation in annotations:
                 annotation['doc_idx'] = i
-        for true_annotation in true_annotations:
+
+        for true_annotation in true_annotations: # This loop covers all predictions which DO overlap.
```
Comment (lhz1029) [inline finetune/util/metrics.py]: Perhaps a simpler version of these two loops:

```
        for true_annotation in true_annotations: 
            for pred_annotation in predicted_annotations:
                if equality_fn(true_annotation, pred_annotation):
                    if pred_annotation['label'] == true_annotation['label']:
                            d[true_annotation['label']]['true_positives'].append(true_annotation)
                            break
            else:
                d[true_annotation['label']]['false_negatives'].append(true_annotation)

        for pred_annotation in predicted_annotations: 
            for true_annotation in true_annotations:
                if (
                    equality_fn(
                        true_annotation, pred_annotation
                    ) and true_annotation["label"] == pred_annotation["label"]
                ):
                    break
            else:
                d[pred_annotation['label']]['false_positives'].append(pred_annotation)
```

**Review by lhz1029 — APPROVED** [review]

File: finetune/util/metrics.py (hunk 6 above)
Comment (benleetownsend) [reply finetune/util/metrics.py]: Thanks, this is much cleaner.

File: tests/test_metrics.py (hunk 2 above)
Comment (madisonmay) [reply tests/test_metrics.py:119]: I think it's probably good practice to use almostEqual in place of Equal when doing floating point equality checks.

**Pull request merged**